### ISSUE TITLE
Random Freezing Parts in Downloaded Episodes - 3-4 seconds Randomly Missing Data  in Each Download

### ISSUE THREAD

**Issue by 4github4** [hama3254/Crunchyroll-Downloader-v3.0#303]
When I download a CR episode , every time I download video always gets a different random time stamp that video or audio freezes during playback. I can never have a video with no freezes no matter how many times I tried

for example I downloaded same episode 2 times, 

in first download attempt - durations between 16:24-16:27 are corruped or missing data - I cant hear or see video, its freezes

in second download attempt- durations between 12:21-12:25 are corruped or missing data - I cant hear or see video, its freezes

Sometimes some download attems have longer freeze or more, sometimes they have less, but they always have a frozen part in episode. Also video file sizes are 2-3 mb different in each download. So this might also explain some data are randomly lost 

When video parts are muxed together maybe some data are being forgotten/lost or not muxed into final video/audio data. Maybe. 
 can we have option to download m3u8 files to a folder manually to mix it manually?
or a better idea -> Can we have option like ''double-pass'' or slower download so it can make sure none of the data parts are forgotten or lost ?? because as I mentioned everytime I download same episode, different durations or parts of episode are missing. So double pass might be a good idea.   Just some ideas for solution even though I am not good at coding.

**Comment by hama3254** [comment]
That already a known issue :

>That is not an freeze, there is most likely one of the fragments from the video stream just missing. 
This might happen due to any kind of network issue, the problem the default download library i use (ffmpeg) can not retry failed fragments it just skips them.
I have made two attempts to solve this issue.
>
>1. the hybrid mode ( https://www.youtube.com/watch?v=s_SY3RM1F3Y )
the hybrid mode opens the stream file and saves all parts to an temporary directory which allows me to have more control over the download itself, the downside if anything is wrong with the settings it might crash in the build process and you need the storage space to have cached `.ts` fragments on the drive.    
>
>2. the error handling function 
The error handling pauses the download process and give the user the information that the set amount of fragments have failed, you can then kill the download or retry from the beginning or just ignore it (it's not perfect but a 2-10 sek skip is not that bad especially at the end of the download)

_Originally posted by @hama3254 in https://github.com/hama3254/Crunchyroll-Downloader-v3.0/issues/119#issuecomment-751484212_

**Comment by 4github4** [comment]
Thank you so much for the information, yes its more like fragments are missing.

I have one question. where is the .ts files are stored in drive when hybrid mode is used?

**Comment by hama3254** [comment]
> Thank you so much for the information, yes its more like fragments are missing. Now I prefer to use downloader's 3.2 version because new 3.7 one had multiple 2-4 sec missing fragments when I tried. I dont know why.Maybe I should try hybrid mode as you said in 3.7.

The Download itself should not be really different between version since it is normally done by the `ffmpeg.exe` 

> 
>     * I have one question. how can I prevent those .ts files to get deleted??? is there option to keep them or will there?

I would add an option but the files are as download which means AES encrypted so i would also need to add the key into the folder and then you still need something that can read and decrypt them and that is not really easy (that the main reason I started with ffmpeg for processing in the first place)

**Comment by 4github4** [comment]
> I would add an option but the files are as download which means AES encrypted so i would also need to add the key into the folder and then you still need something that can read and decrypt them and that is not really easy (that the main reason I started with ffmpeg for processing in the first place)

 Yess I copied half of the files to another folder while it still downloads, but I couldnt open the part ts files in VLC, because they are not readable. 
Actually it would be great to add that option even though not everyone can use it
I dont know how to use commandline so I dont how how to use key and decrypt and merge files by ffmpeg.  Maybe I can learn how to use commandline ffmpeg ts merging process with key or  I might find a GUI version of ffmpeg ...                                  
so maybe if u add the option to keep ts files and also the key somehow, it would still be good.

**Comment by hama3254** [comment]
>  but I couldnt open the part ts files in VLC, because they are not readable.

well you did not have a key since it needs to be in the same folder as ffmpeg because it just don't find the key file if the path is local which however is fine for the video files so ....
The good new is i added a popup after you enable the hybrid mode which gives you the choice to keep the temp files which does not only keep the files but also renames the folder like the mp4 so that you know that's inside and a second m3u8 playlist gets created for VLC and of course a copy of the key file is also in that folder now. 


> I dont know how to use commandline so I dont how how to use key and decrypt and merge files by ffmpeg. Maybe I can learn how to use commandline ffmpeg ts merging process with key or I might find a GUI version of ffmpeg ...

You could try VLC for that, it has an UI and works ok. [A quick Video how it work is here](https://www.youtube.com/watch?v=miUL5lvLQnk)

The "Old second idea" is not really necessary since the downloader already does one retry for failed downloads anyway and you could download fragments that failed twice manually because i just added the url to the and the path to the error console just look out for `"Download error #2: "` if you miss something 

You should find it in the console which can be displayed clicking the grey bar on the very bottom of the downloader

> ![grafik](https://user-images.githubusercontent.com/52826614/119235080-01a03400-bb31-11eb-9e79-5c79049482bb.png)

**Comment by hama3254** [comment]
i released update 10 for the 3.8-Beta 

https://github.com/hama3254/Crunchyroll-Downloader-v3.0/releases/tag/v3.8-Beta

**Comment by 4github4** [comment]
Thank you so much. I saw the new popup in 3.8 and its greatt. I will try it soon nowww. Thank you so much for considering my feed back and making downloader more perfect :))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))) Its so kind of you

And, yess I will try vlc, I didnt know it can mux fragment files. Thank you I watched ur video and give it a like and comment

You are right I didnt try that error feature yet. Now I understand it more. I didnt know that console . U really helped me. I will try  it out soon . And Sorry for my late understanding ( because I didnt know that it was like what I kept calling double pass : D)
